DOC Add HGBDT to "user_guide" reference in RF

amay1212 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Addresses comment number 2 from @ArturoAmorQ in issue #26220

#### What does this implement/fix? Explain your changes.
This updates the documentation of HGBT in RF section

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

amay1212 committed:
HGBT Doc added to "user_guide" reference in RF

amay1212 commented:
> Thanks for addressing this last point in the issue :) here are a few comments:
> 
> * I think the paragraph should be added at the end of the [Random Forests subsection](https://scikit-learn.org/stable/modules/ensemble.html#random-forests) as it does not concern the other subsections in the [Forests of randomized trees](https://scikit-learn.org/stable/modules/ensemble.html#forests-of-randomized-trees)
> * The text is almost an exact copy-paste of the note in the [Gradient Tree Boosting section](https://scikit-learn.org/stable/modules/ensemble.html#gradient-tree-boosting), which is already redundant with the info in the [Histogram-base Gradient Boosting section](https://scikit-learn.org/stable/modules/ensemble.html#histogram-based-gradient-boosting). Repeating said text does not bring any new information regarding RFs estimators themselves.
> * The statement on being faster is not always true, as RFs can be parallelized using `n_jobs`. Overall, the performance of HGBTs versus parallelized RFs depends on the specific characteristics of the dataset and the modeling task. It's always a good idea to try both models and compare their performance on your specific problem to determine which model is the best fit.
> * Not a strong opinion about this but I'd rather not use a note, as they are meant to be brief and here we want to give a deeper insight.
> 
> Please let me know if you need help or more details addressing my comments.

@ArturoAmorQ 
Thanks for the feedback, I believe that's a great point about how performance can be increased in RF as well. I think the first two points are quite straightforward. However, for the last two points, I have following assumptions

- I believe you want me to include code examples for both HGBTs and RFs, so that it is clear how the models are implemented and how their speed can be compared.
- For the last point, do you want me to add the comments or description in **section** or **subtitle**, if you could confirm.

ArturoAmorQ commented:
It is not necessary to add code. For that purpose I opened https://github.com/scikit-learn/scikit-learn/pull/26320. What I had in mind is comparing the computational cost of the algorithms themself, for instance, building shallow (hgbt) vs deep (rf) trees. Mention that in rf trees are independent and can be built separately whereas hgbt is the corrections are built succesively. Mention the efficiency of binning, etc.

And for the last point I meant plain text, just not the note formatting. 

Please let me know if it's clearer now.

amay1212 commented:
> It is not necessary to add code. For that purpose I opened https://github.com/scikit-learn/scikit-learn/pull/26320. What I had in mind is comparing the computational cost of the algorithms themself, for instance, building shallow (hgbt) vs deep (rf) trees. Mention that in rf trees are independent and can be built separately whereas hgbt is the corrections are built succesively. Mention the efficiency of binning, etc.
> 
> And for the last point I meant plain text, just not the note formatting. 
> 
> Please let me know if it's clearer now.

@ArturoAmorQ 
Sure, It's clear now. However, just a suggestion to include a separate title as mentioned above to make it more clearer and noticeable. Let me know if it makes sense.

amay1212 committed:
#issue:#26220  Made changes as per review comments

amay1212 commented:
> > It is not necessary to add code. For that purpose I opened #26320. What I had in mind is comparing the computational cost of the algorithms themself, for instance, building shallow (hgbt) vs deep (rf) trees. Mention that in rf trees are independent and can be built separately whereas hgbt is the corrections are built succesively. Mention the efficiency of binning, etc.
> > And for the last point I meant plain text, just not the note formatting.
> > Please let me know if it's clearer now.
> 
> @ArturoAmorQ 
Thanks for the quick feedback!
It's clear now. Also, I have made the changes and addressed all the points mentioned in the review comments. Please review and let me know in case of any concerns further.

ArturoAmorQ commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,16 @@ In contrast to the original publication [B2001]_, the scikit-learn
+
+   - Sequential boosting: In HGBT, the trees are built sequentially, with each tree correcting the mistakes of the previous trees. This can be more computationally efficient than RF, where all the trees are built independently and in parallel.
+
+   - Efficient binning: HGBT uses an efficient binning algorithm that can handle large datasets with a high number of features. The binning algorithm can pre-process the data to speed up the subsequent tree construction. In contrast, RF does not use binning and relies on randomly selecting features at each split, which can be computationally expensive.
```
Here `XXX` is to be filled by you with a reference to the [Why it's faster section](https://scikit-learn.org/stable/modules/ensemble.html#why-it-s-faster).

```suggestion
- Efficient binning: HGBT uses an efficient binning algorithm that can handle
  large datasets with a high number of features. The binning algorithm can
  pre-process the data to speed up the subsequent tree construction (see
  subsection XXX). In contrast, the scikit-learn implementation of random
  forests does not use binning and relies on exact splitting, which can be
  computationally expensive.
```

amay1212 replied:
> Here `XXX` is to be filled by you with a reference to the [Why it's faster section](https://scikit-learn.org/stable/modules/ensemble.html#why-it-s-faster).

I have applied the changes. Please review

ArturoAmorQ reviewed (commented):
Another comment: Please try to keep lines below 80 characters to comply with the [PEP 8 style convention](https://peps.python.org/pep-0008/#maximum-line-length)

amay1212 committed:
Added changes as per the review comments

amay1212 commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,16 @@ In contrast to the original publication [B2001]_, the scikit-learn
 
+Comparison of Histogram-based Gradient Boosting Trees over Random Forests in terms of computational cost:
+
+   - Building shallow trees: HGBT can be computationally more efficient than RF when building shallow trees because it only needs to consider a limited number of bins to construct the splits, whereas RF builds deep trees that require more iterations to reach optimal splits.
```
I have applied the changes. Please review

amay1212 commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,16 @@ In contrast to the original publication [B2001]_, the scikit-learn
 implementation combines classifiers by averaging their probabilistic
 prediction, instead of letting each classifier vote for a single class.
 
+Comparison of Histogram-based Gradient Boosting Trees over Random Forests in terms of computational cost:
```
> Instead of a subtitle, here you can introduce the goal of the following bullet-points like this:

Made the changes as requested.

amay1212 commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,16 @@ In contrast to the original publication [B2001]_, the scikit-learn
+
+   - Building shallow trees: HGBT can be computationally more efficient than RF when building shallow trees because it only needs to consider a limited number of bins to construct the splits, whereas RF builds deep trees that require more iterations to reach optimal splits.
+
+   - Sequential boosting: In HGBT, the trees are built sequentially, with each tree correcting the mistakes of the previous trees. This can be more computationally efficient than RF, where all the trees are built independently and in parallel.
```
> The phrase "This can be more computationally efficient than RF" is a bit vague, I suggest giving a deeper explanation on this regard:

Made the changes as requested

amay1212 commented:
> Another comment: Please try to keep lines below 80 characters to comply with the [PEP 8 style convention](https://peps.python.org/pep-0008/#maximum-line-length)

Thanks for noticing it!
Just made the changes, let me know if there are any concerns.

adrinjalali commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,19 @@ In contrast to the original publication [B2001]_, the scikit-learn
+A competitive alternative to random forests are
+:ref:`histogram_based_gradient_boosting` (HGBT) models:
+
+- Building trees: Random forests typically rely on deep trees (that overfit individually) which uses much computational resources, as they require several splittings and evaluations of candidate splits. Boosting models build shallow trees (that underfit individually) which are faster to fit and predict.
```
lines too long, please limit them to 88 chars like the rest of the document.

adrinjalali reviewed (commented):
This is a good addition. I wonder if we want to refer to xgboost here or not, since that's what people know.

ArturoAmorQ commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,19 @@ In contrast to the original publication [B2001]_, the scikit-learn
+
+  In contrast, random forests (RF) use an ensemble of decision trees, where each tree is built independently and is not influenced by the other trees. The final prediction is made by taking the majority vote of all the trees in the ensemble. While RF can achieve high accuracy, they may require a larger number of trees to achieve the same level of accuracy as HGBT. This is because RF may not be ab ...
+
+- Efficient binning: Efficient binning: HGBT uses an efficient binning algorithm that can handle large datasets with a high number of features. The binning algorithm can pre-process the data to speed up the subsequent tree construction (see `Why-it's-faster <https://scikit-learn.org/stable/modules/ensemble.html#why-it-s-faster>`__). In contrast, the scikit-learn implementation of random forests d ...
```
For the reference, sphinx supports reStructured text:

https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html

amay1212 replied:
Sure, made the changes.

amay1212 committed:
Updated line formatting changes and used cross reference links instead of normal url's

adrinjalali commented on doc/modules/ensemble.rst:
```diff
@@ -1211,7 +1251,7 @@ means that ``[{0}]`` is equivalent to ``[{0}, {1, 2}]``.
      <10.3390/jrfm15050193>`.
      Journal of Risk and Financial Management 15, no. 5: 193
-
+     
```
remove whitespace

adrinjalali reviewed (approved):
This is certainly an improvement.

amay1212 committed:
Removed unnecessary whitespaces

ArturoAmorQ commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,46 @@ In contrast to the original publication [B2001]_, the scikit-learn
+   build shallow trees (that underfit individually) which are faster to fit
+   and predict.
+
+-  Sequential boosting: In histogram-based gradient boosting (HGBT), the
+   decision trees are built sequentially, where each tree is trained to
+   correct the errors made by the previous trees. This is achieved by
+   assigning higher weights to the training instances that were misclassified
+   in the previous iteration, which helps the next tree to focus on those
+   instances and improve their classification. This process is repeated
+   iteratively until a predefined number of trees are built or until the
+   model's performance stops improving.
+
+   In contrast, random forests (RF) use an ensemble of decision trees, where
+   each tree is built independently and is not influenced by the other trees.
+   The final prediction is made by taking the majority vote of all the trees
+   in the ensemble. While RF can achieve high accuracy, they may require a
+   larger number of trees to achieve the same level of accuracy as HGBT. This
+   is because RF may not be able to correct errors made by previous trees as
+   effectively as HGBT, which can lead to a larger number of trees needed to
+   achieve the desired level of accuracy.
```
I would say that explaining each method in depth is distracting, it is better to keep a short paragraph focusing on the differences and their consequences. Also notices that the abbreviation "HGBT" was already introduced above, so there is no need to redefine it.

```suggestion
-  Sequential boosting: In HGBT, the decision trees are built sequentially,
   where each tree is trained to correct the errors made by the previous ones.
   This allows them to iteratively improve the model's performance using
   relatively few trees. In contrast, random forests use a majority vote to
   predict the outcome, which can require a larger number of trees to achieve
   the same level of accuracy.
```

ArturoAmorQ reviewed (commented):
Please notice that you can accept suggestions directly from github. This makes the review process easier.

amay1212 committed:
Reformatted  Sequential Boosting information

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

amay1212 commented on doc/modules/ensemble.rst:
```diff
@@ -150,6 +150,46 @@ In contrast to the original publication [B2001]_, the scikit-learn
+   effectively as HGBT, which can lead to a larger number of trees needed to
+   achieve the desired level of accuracy.
+
+-  Efficient binning: Efficient binning: HGBT uses an efficient binning
```
Done!
Thanks for reviewing

amay1212 commented:
> Please notice that you can accept suggestions directly from github. This makes the review process easier.

Sure, will keep in mind

ArturoAmorQ reviewed (approved):
Thanks for your effort @amay1212! It certainly LGTM :)
